bleak/__init__.py

```diff
@@ -204,6 +205,52 @@ def set_scanning_filter(self, **kwargs):
         )
         self._backend.set_scanning_filter(**kwargs)
 
+    async def devices(self) -> AsyncIterator[Tuple[BLEDevice, AdvertisementData]]:
+        """
+        Yields devices and their advertising data as they are discovered.
+
+        .. note::
+            Ensure that scanning is started before calling this method.
```

Comment:
I think it is also important to note that there must not be any `await` between starting scanning and calling `devices()` otherwise advertisements can be missed. (This pitfall was my main reason for preferring the proposed `scan()` over `devices()` until I realized the non-deterministic cleanup issues with `scan()`).

It would be very helpful to include an example snippet here in the documentation that shows proper usage.

Reply:
Actually, I was aware of this, and it was kinda deliberate. I create the queue when the users starts iterating and destroy it afterward.
If I create the queue in the `__init__()` and add devices to it... What if the user never uses the async iterator, what if they just start the scanning and do nothing? When the queue fills up, should I remove the first found device and add the new one, or just stop adding new devices?

I actually don't see any drawback in skipping few advertisements... After all, what if the user started the scan a few moments later?

As I am mostly using the `BleakScanner()` as an instance instead of using the class methods, I wanted to allow user the flexibility: create the scanner; use the callback for a while, if you want; remove the callback; use it as an iterator; etc.. Unusual use case, but why not.

Now when you put it like that, I am thinking... But maybe the answer lies in the docstring themselves:
- `discovered_devices()`: Gets list of the devices that the scanner **has discovered during the scanning**.
I agree that this shall return all devices since `.start()`, and it does.
- `devices()`: Yields devices and their advertising data **as they are discovered**.
This just yields any device discovered [only] while you are using this method (iterating it).

Maybe I should really rename the `devices()` method. You suggested `watch_advertisements`, which seems kinda long, and everything else is based around the notion of _devices_. I would like to avoid `scan` (for reasons explained at the end of https://github.com/hbldh/bleak/pull/1361#discussion_r1258890610 ). `observe_devices` seems like we are observing their state or something...

Reply:
> I actually don't see any drawback in skipping few advertisements... After all, what if the user started the scan a few moments later?

Agreed.

> Maybe I should really rename the `devices()` method.

I like having `advertisements` in the name since each iteration is triggered by new advertisement data rather than a new device. I think `observe` is a good name since it fits with BLE broadcast/observe terminology.

Reply:
Ok, I actually asked global buzzword for a help

> The distinction between discovered_devices() and devices() is indeed a bit subtle given their current names. Here are a few alternative names for the devices() method that might make this distinction clearer:
>- `live_devices()`: This suggests that the method provides a real-time feed of devices as they are discovered.
>- `streaming_devices()`: This also suggests a continuous feed of devices, and aligns with terms like "streaming data".
>- `dynamic_devices()`: This implies that the list of devices can change over time, which is the case for this method.
>- `discover_live_devices()`: This combines elements of a couple of the above suggestions.
>
> Remember, the key is to find a name that makes it clear this method is different from `discovered_devices()` in that it provides a live, ongoing feed of discovered devices.

> More suggestions please. You can also leave out `devices` word if you come up with any good alternative.

> Sure, here are a few more suggestions, dropping the devices word for some alternatives:
>- `monitor()`: This implies that the method is actively monitoring for new devices.
>- `discover_live()`: This suggests that the method is discovering devices in real-time.
>- `detect()`: This could imply a continuous process of detection.
>- `observe()`: This also suggests an ongoing, active process.
>- `watch()`: This is short, and implies a continuous process.
>- `continuous_discovery()`: A bit longer, but very descriptive.
>- `stream_discoveries()`: Suggests that this is an ongoing, streaming process.
>- `find_as_available()`: Implies that devices are yielded as soon as they're found.

I still can't decide. I kiiind of like `monitor` or `observe`, but the most fitting names are already taken :smile:

Reply:
Renamed to `observe()`

Reply:
I feel like since `observe()` is a verb, it makes it sound like this is starting scanning. How about `observed_changes()` instead? I haven't really encountered any many async generators yet, so I don't know if there is a common naming pattern of what feels the most Pythonic (the ones I have used were binding a C++ library, so they don't really count).

```python
async with BleakScanner() as scanner:
    async for dev, ad in scanner.observed_changes():
```

Another reason for this name is that after thinking about it some more this is more descriptive of how the OS Bluetooth stacks actually work. We don't necessarily receive an event for every single advertisement packet, but rather receive an event when something changes, either the RSSI or the value of one of the data. It would be helpful to include a variation of the previous sentence in the docstring.

Reply:
> We don't necessarily receive an event for every single advertisement packet, but rather receive an event when something changes

Sometimes getting each advertisement packet is a desired behavior, but hard to achieve on certain OS (e.g. my tries with passive scanning and `AdvertisementMonitor` in the past). I would say that one wants to either _find a BLE device_, or _scan advertisement packets_ – filtering later to only get packets where something changes shall be an option/feature, not a limitation (however this is domain of OS BT stack, not `bleak`).

> I haven't really encountered any many async generators yet, so I don't know if there is a common naming pattern of what feels the most Pythonic

I haven't seen or found any naming convention on this topic. Looking at some existing methods, it seems that the name represents what the method _is doing_ internally (verb), if anything, or what it represents (noun):
- `enumerate()`, `zip()`, `map()`, `filter()`, `walk()`, `scandir()`
- `keys()`, `values()`, `items()`, `union()`, `lines()`, `nodes()`, `children()`, `elements()`, `records()`

I still somehow see this method having almost the same functionality as `discovered_devices_and_advertisement_data` property, only that the devices are yielded in real-time, not after the scan, and there are duplicates.

If we ignore all async timeout related overhead in the example below, they are indeed very similar:
```python
    async with BleakScanner() as scanner:
        await asyncio.sleep(1)
    print("Devices:", scanner.discovered_devices_and_advertisement_data)

    # The same, except it prints all advertisement packets instead of devices
    async with BleakScanner() as scanner:
        found = []
        try:
            async with async_timeout(1):
                async for bd_ad in scanner.observe():
                    found.append(bd_ad)
        except asyncio.TimeoutError:
            pass
    print("Adv. packets:", found)
```
With this logic, I could use `discovered_devices_and_advertisement_data` as a starting point:
1. Remove `discovered_`, because this method works in the real time instead of reporting the past.
**\>** `devices_and_advertisement_data`
2. The resulting name is too long, and it exists only for backward compatibility – if we could, we would have just `discovered_devices` which would return the same.
**\>** `devices`
3. `devices` represents the actual `BLEDevice`s, created from, and updated with `AdvertisementData` during scanning. This is not true for this method, as it returns each `AdvertisementData` packet, and `BLEDevice` associated with (created from) it.
**\>** `advertisements`?
**\>** `advertisement_data`?

Reply:
> 3\. `advertisement_data`

I think this is the best of all the ideas we have come up with so far.